[RISCV] Add Qualcomn uC Xqcili (load large immediates) extension

u4f3 opened the pull request:
The Xqcili extension includes a two instructions that load large immediates than is available with the base RISC-V ISA.

The current spec can be found at: https://github.com/quic/riscv-unified-db/releases/tag/Xqci-0.7.0

This patch adds assembler only support.

topperc reviewed (approved):
LGTM

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoXqci.td:
```diff
@@ -588,6 +590,19 @@ let Predicates = [HasVendorXqcilo, IsRV32] in {
 
+let Predicates = [HasVendorXqcili, IsRV32] in {
+let hasSideEffects = 0, mayLoad = 0, mayStore = 0 in {
+  def QC_LI : RVInstU<RISCVOpcode<"QC_LI",0b0011011>, (outs GPRNoX0:$rd), (ins simm20:$imm20),
```
Wrap this line to keep it under 80 characters.

svs-quic replied:
This can be OPC_OP_IMM_32?

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoXqci.td:
```diff
@@ -588,6 +590,19 @@ let Predicates = [HasVendorXqcilo, IsRV32] in {
+let Predicates = [HasVendorXqcili, IsRV32] in {
+let hasSideEffects = 0, mayLoad = 0, mayStore = 0 in {
+  def QC_LI : RVInstU<RISCVOpcode<"QC_LI",0b0011011>, (outs GPRNoX0:$rd), (ins simm20:$imm20),
+                      "qc.li", "$rd, $imm20">{
```
Space before {

svs-quic commented on llvm/lib/Target/RISCV/RISCVFeatures.td:
```diff
@@ -1358,6 +1358,12 @@ def HasVendorXqciint
       AssemblerPredicate<(all_of FeatureVendorXqciint),
                          "'Xqciint' (Qualcomm uC Interrupts Extension)">;
 
+def FeatureVendorXqcili : RISCVExperimentalExtension<0, 2, "Qualcomm uC Load Large Immediate Extension", []>;
```
This needs [FeatureStdExtZca] as well given that it has a 48 bit instruction.

u4f3 replied:
It seems that this requirement is not explicitly specified in the spec, or I'm missing some part of documents somewhere. Anyway I add [FeatureStdExtZca].

svs-quic reviewed (commented):
Please add entries in RISCVUsage.rst and ReleaseNotes.md.

topperc reviewed (changes requested):
Dropping my approval based on other comments.

hchandel commented on llvm/test/CodeGen/RISCV/attributes.ll:
```diff
@@ -412,6 +413,7 @@
 ; RV32XQCICS: .attribute 5, "rv32i2p1_xqcics0p2"
 ; RV32XQCICSR: .attribute 5, "rv32i2p1_xqcicsr0p2"
 ; RV32XQCIINT: .attribute 5, "rv32i2p1_zca1p0_xqciint0p2"
+; RV32XQCILI: .attribute 5, "rv32i2p1_xqcili0p2"
```
Enabling ZCA will require replacing `rv32i2pl_xqcili0p2` with `rv32i2pl_zca1p0_xqcili0p2`

u4f3 committed:
[RISCV] Add Qualcomn uC Xqcili (load large immediates) extension

The Xqcili extension includes a two instructions that load large immediates than is available with the base RISC-V ISA.

The current spec can be found at: https://github.com/quic/riscv-unified-db/releases/tag/Xqci-0.7.0

This patch adds assembler only support.

u4f3 commented:
Seems that the error in `/llvm-project/clang/test/Driver/offload-Xarch.c` is not relevant. And I can't reproduce this error since I don't have a windows machine...

lenary commented:
> Seems that the error in `/llvm-project/clang/test/Driver/offload-Xarch.c` is not relevant. And I can't reproduce this error since I don't have a windows machine...

Yeah, don't worry about that error.

svs-quic commented on llvm/test/MC/RISCV/xqcili-invalid.s:
```diff
@@ -0,0 +1,21 @@
+# Xqcili - Qualcomm uC Load Large Immediate Extension
+# RUN: not llvm-mc -triple riscv32 -mattr=+experimental-xqcili < %s 2>&1 \
```
Please add a run line with mattr=-experimental-xqcili

svs-quic reviewed (approved):
LGTM! @lenary do you have any comments?

lenary reviewed (approved):
LGTM. Thanks!

svs-quic reviewed (approved):
LGTM!

u4f3 commented:
Rebased to solve conflicts. Please merge it for me when you think this is ready.

topperc reviewed (approved):
LGTM

lenary reviewed (approved):
LGTM.

Thank you for doing this PR for Xqcili support, it's great to see interest in this from the RISC-V community beyond Qualcomm. In future, it would be great to coordinate on what you are planning to upstream, as we (Qualcomm) have downstream implementations of the extensions which we are in the process of upstreaming, and we would like to avoid duplicated work in this area.
